{
	"id": "GHSA-x9xc-63hg-vcfq",
	"advisories": [
		{
			"id": "GHSA-x9xc-63hg-vcfq",
			"title": "cassandra-rs's non-idiomatic use of iterators leads to use after free",
			"description": "### Impact\nCode that attempts to use an item (e.g., a row) returned by an iterator after the iterator has advanced to the next item will be accessing freed memory and experience undefined behaviour. Code that uses the item and then advances the iterator is unaffected. This problem has always existed.\n\nThis is a use-after-free bug, so it's rated high severity. If your code uses a pre-3.0.0 version of cassandra-rs, and uses an item returned by a cassandra-rs iterator after calling `next()` on that iterator, then it is vulnerable. However, such code will almost always fail immediately - so we believe it is unlikely that any code using this pattern would have reached production. For peace of mind, we recommend you upgrade anyway.\n\n### Patches\nThe problem has been fixed in version 3.0.0. Users should upgrade to ensure their code cannot use the problematic pattern.\n\n### Workarounds\nEnsure all usage fits the expected pattern. For example, use `get_first_row()` rather than an iterator, or completely process an item before advancing the iterator with `next()`.\n\n### References\nNone.\n",
			"severity": [
				{
					"type": "cvss_v31",
					"source": "github.com/advisories",
					"cvss_v31": {
						"vector": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H",
						"base_score": 7.5,
						"base_severity": "HIGH",
						"temporal_score": 7.5,
						"temporal_severity": "HIGH",
						"environmental_score": 7.5,
						"environmental_severity": "HIGH"
					}
				},
				{
					"type": "vendor",
					"source": "github.com/advisories",
					"vendor": "HIGH"
				}
			],
			"references": [
				{
					"source": "github.com/advisories",
					"url": "https://github.com/Metaswitch/cassandra-rs/security/advisories/GHSA-x9xc-63hg-vcfq"
				},
				{
					"source": "github.com/advisories",
					"url": "https://nvd.nist.gov/vuln/detail/CVE-2024-27284"
				},
				{
					"source": "github.com/advisories",
					"url": "https://github.com/Metaswitch/cassandra-rs/commit/ae054dc8044eac9c2c7ae2b1ab154b53ca7f8df7"
				},
				{
					"source": "github.com/advisories",
					"url": "https://github.com/Metaswitch/cassandra-rs"
				},
				{
					"source": "github.com/advisories",
					"url": "https://rustsec.org/advisories/RUSTSEC-2024-0017.html"
				}
			],
			"published": "2024-04-05T15:00:55Z",
			"modified": "2024-04-05T15:00:55Z",
			"optional": {
				"github_reviewed": true
			}
		}
	],
	"vulnerabilities": [
		{
			"id": "CVE-2024-27284"
		}
	],
	"detection": [
		{
			"ecosystem": "cargo",
			"vulnerable": true,
			"package": {
				"name": "cassandra-cpp"
			},
			"affected": {
				"type": "semver",
				"range": [
					{
						"lt": "3.0.0"
					}
				],
				"fixed": [
					"3.0.0"
				]
			}
		}
	],
	"data_source": "cargo-ghsa"
}
